go-mode.el

```diff
@@ -1562,6 +1568,185 @@ for."
       (if (not (eq cur-buffer (current-buffer)))
           (display-buffer (current-buffer) `(,go-coverage-display-buffer-func))))))
 
+(defun go-goto-function ()
+  "Go to the function defintion above point.
+
+If we are on a docstring, follow the docstring down.
+If no function is found, assume that we are at the top of a file
+and search forward instead.
+
+If we are inside an anonymous function, go to that.
+If we are inside a function that has an anonymous function inside of it and we
+are below that anonymous function, go to the root function."
+  (interactive)
+  (let ((p (point)))
+    (cond
+     ((save-excursion
+        (beginning-of-line)
+        (looking-at "^//"))
+      ;; In case we are looking at the docstring, move on forward until we are
+      ;; not anymore
+      (beginning-of-line)
+      (while (looking-at "^//")
+        (forward-line 1))
+      ;; If we are still not looking at a function, retry by calling self again.
+      (when (not (looking-at "^func"))
+        (go-goto-function)))
+
+     ((not (looking-at "^func"))
+      ;; If we are not looking at the beginning of a function line, do a regexp
+      ;; search backwards
+      (re-search-backward "\\<func\\>" nil t)
+
+      ;; If nothing is found, assume that we are at the top of the file and
+      ;; should search forward instead.
+      (when (not (looking-at "func"))
+        (re-search-forward "\\<func\\>" nil t)
+        (forward-word -1))
+
+      ;; If we have landed at an anonymous function, it is possible that we
+      ;; were not inside it but below it. If we were not inside it, we should
+      ;; go to the containing function.
+      (while (go--in-function-p p)
+        (go-goto-function)))))
+
+  ;; If we are still in a comment, redo the call so that we get out of it.
+  (when (go-in-comment-p)
+    (go-goto-function)))
+
+(defun go--in-function-p (compare-point)
+  "Return t if point is inside the function that starts at `compare-point', nil
+otherwise."
+  ;; Check that we are not looking at a top level function. If we are,
+  (when (not (looking-at "^func"))
+   (save-excursion
+     (go--goto-return-values)
+     ;; Try to place the point on the opening brace
+     (cond
+      ((looking-at "(")
+       (forward-list 1)
+       (forward-char 1))
+      ((not (looking-at "{"))
+       (forward-word 1)
+       (forward-char 1)))
+
+     (when (looking-at "{")
+       ;; Go past the body of the function and back inside of it.
+       (forward-list 1)
+       (backward-char 1)
+       ;; Finally, compare if our position is past the assert point.
```

Comment:
This comment confuses me. Your code checks if `point < compare-point` – how is that being "past the assert point"?